```{r}
library(phyloseq)
library(ggplot2)
library(microbiome)
library(microbiomeutilities)
library(ranacapa)
library(plotly)

library(microViz)
library(ggpubr)
library(htmlwidgets)

library(ANCOMBC)
library(Maaslin2)

library(cols4all)
```

```{r}
# metadata
meta_df <- read.csv("data/alternative/total.csv", sep = ",", header = TRUE)
rownames(meta_df) <- meta_df$run.ID
meta_df$associated.phenotype <- as.factor(meta_df$associated.phenotype)
meta_df$host.age <- as.integer(meta_df$host.age)
meta_df$BMI <- as.integer(meta_df$BMI)
head(meta.df)

summary(meta_df$associated.phenotype)
summary(meta_df$host.age)
summary(meta_df$BMI)
meta_df <- meta_df[,c("project.ID", "run.ID", "associated.phenotype")]

sampledata <- sample_data(meta_df)
sample_names(sampledata) <- meta_df$run.ID
head(sampledata)

# count table
abs_count_df <- read.csv(
  "data/alternative/absolute_count_raw.csv",
  sep = "\t",
  header = TRUE
)
rownames(abs_count_df) <- abs_count_df$query_name
abs_count_df <- subset(abs_count_df, select = -query_name)
otu <- otu_table(as.matrix(t(abs_count_df)), taxa_are_rows = TRUE)

# asv table
taxa_df <- read.csv("data/alternative/tax_table.csv", sep = "\t", header = TRUE)
rownames(taxa_df) <- taxa_df$Ident
taxa_df <- subset(taxa_df, select = -c(Ident, X, Gtdb_representative))
head(taxa_df)

tax <- tax_table(as.matrix(taxa_df))
```

```{r}
physeq <- phyloseq(otu, tax, sampledata)
```

# Remove singleton
```{r}
#sort sample total reads, prune taxa
all.ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(physeq) > 0, physeq)

#explore singleton
all.tx.pres <- rowSums(abundances(all.ps) > 0)
all.singleton <- as.vector(
  as.matrix(
    tax_table(all.ps)[names(which(all.tx.pres == 1)), "Genus"]
  )
)
all.singleton #793 singleton taxa

names(all.singleton) <- rownames(
  tax_table(all.ps)[names(which(all.tx.pres == 1)), "Genus"]
)
sort(table(all.singleton),
     decreasing = TRUE)

## most common singleton taxa are

#counts of each singleton taxa
sort(taxa_sums(all.ps)[names(all.singleton)], decreasing = TRUE)
summary(taxa_sums(all.ps)[names(all.singleton)]) #median 227 // 3rd quartile 642

table(
  all.singleton[names(which(taxa_sums(all.ps)[names(all.singleton)] < 642))]
)
all.taxa.rm <- names(which(taxa_sums(all.ps)[names(all.singleton)] < 642))

## Remove singleton taxa that have read abundance < 515 reads
all.fil <- prune_taxa(!(taxa_names(all.ps) %in% all.taxa.rm), all.ps)
all.fil #4326 taxa
summary(sample_sums(all.fil) / sample_sums(all.ps)) #median 100%

all.ggrare <- ggrare(
  all.fil,
  step = 100,
  label = NULL,
  color = "associated.phenotype"
)
all.ggrare.plot <- all.ggrare +
  # xlim(0, 3000) +
  theme(legend.position = "bottom") +
  theme_classic()

all.ggrare.plot

sort(sample_sums(all.fil))
```

# PCoA
```{r}
ord <- ordinate(all.fil, "PCoA", "bray")

#Plot scree plot
phyloseq::plot_scree(ord) +
  geom_bar(stat = "identity", fill = "firebrick") +
  scale_x_discrete(limits = c(1:10)) +
  labs(x = "Axis", y = "Proportion of Variance")

plot_ordination(all.fil, ord,
  type = "samples",
  shape = "associated.phenotype",
  color = "project.ID"
)
```

# A-div
```{r}
adiv <- data.frame(
  "Shannon" = phyloseq::estimate_richness(all.fil, measures = "Shannon"),
  "Simpson" = phyloseq::estimate_richness(all.fil, measures = "Simpson"),
  "Chao1" = phyloseq::estimate_richness(all.fil, measures = "Chao1"),
  "Age" = sample_data(all.fil)$host.age,
  "Disease" = sample_data(all.fil)$associated.phenotype
)

adiv <- subset(adiv, select = -Chao1.se.chao1)
names(adiv)[3] <- "Chao1"

adiv$Disease <- as.factor(adiv$Disease)
adiv$Age <- as.factor(adiv$Age)

# plot
comps <- make_pairs(sample_data(all.fil)$associated.phenotype)

ggplot(adiv, aes(Disease, Shannon)) +
  geom_violin() +
  geom_boxplot(width = 0.5) +
  # geom_jitter() +
  stat_compare_means(
    comparisons = comps,
    method = "wilcox.test",
    paired = FALSE
  )
```

# Tukey post hoc test
```{r}
adiv
#Shannon
shan_div <- adiv[, c("Shannon", "Disease")]
tukey_shan <- TukeyHSD(aov(lm(Shannon ~ Disease,
                              shan_div)))$Disease
tukey_shan <- as.data.frame(tukey_shan)

tukey_shan[order(tukey_shan$`p adj`), ]

#Simpson
simp_div <- adiv[, c("Simpson", "Disease")]
tukey_simp <- TukeyHSD(aov(lm(Simpson ~ Disease,
                              simp_div)))$Disease
tukey_simp <- as.data.frame(tukey_simp)

tukey_simp[order(tukey_simp$`p adj`), ]

#Chao1
chao1_div <- adiv[, c("Chao1", "Disease")]
tukey_chao1 <- TukeyHSD(aov(lm(Chao1 ~ Disease,
                               chao1_div)))$Disease
tukey_chao1 <- as.data.frame(tukey_chao1)

tukey_chao1[order(tukey_chao1$`p adj`), ]
```

```{r}
# standard error
se <- function(x) sd(x) / sqrt(length(x))
```

# Phylum
```{r}
ps_phylum <- microbiome::aggregate_rare(all.fil,
  level = "Phylum",
  detection = 0.1,
  prevalence = 0
)

abund_df <- microbiomeutilities::phy_to_ldf(
  ps_phylum,
  transform.counts = "compositional"
)

ggplot(abund_df, aes(x = Sam_rep, y = Abundance, fill = Phylum)) +
  geom_bar(position = "fill",
    stat = "identity"
  )



top_phylum <- abund_df %>%
  group_by(Phylum) %>%
  summarise(mean_abundance = mean(Abundance),
            se = se(Abundance),
            median = median(Abundance),
            q1 = quantile(Abundance)[2],
            q3 = quantile(Abundance)[4]) %>%
  ungroup()

top_phylum <- top_phylum %>% arrange(desc(mean_abundance))
top_phylum <- as.data.frame(top_phylum)
top_phyla <- head(top_phylum$Phylum, n = 10)


abund_df <- abund_df %>%
  dplyr::mutate(pick_Phylum = ifelse(Phylum %in%
                                       top_phyla,
                                     yes = Phylum,
                                     "Others"))

# Plot
abund_df$pick_Phylum <- as.factor(abund_df$pick_Phylum)
abund_df$pick_Phylum <- factor(abund_df$pick_Phylum,
  levels = c(top_phyla, "Others")
)

abund_df <- abund_df[order(abund_df$host.age, decreasing = FALSE), ]

abund_plot <-
  ggplot(abund_df, aes(x = Sam_rep, y = Abundance, fill = pick_Phylum)) +
  geom_bar(position = "fill", stat = "identity") +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

# Genus
```{r}
ps_genus <- microbiome::aggregate_rare(all.fil,
  level = "Genus",
  detection = 0.1,
  prevalence = 0
)

abund_df <- microbiomeutilities::phy_to_ldf(
  ps_genus,
  transform.counts = "compositional"
)

top_genus <- abund_df %>%
  group_by(Genus) %>%
  summarise(mean_abundance = mean(Abundance),
            se = se(Abundance),
            median = median(Abundance),
            q1 = quantile(Abundance)[2],
            q3 = quantile(Abundance)[4]) %>%
  ungroup()

top_genus <- top_genus %>% arrange(desc(mean_abundance))
top_genus <- as.data.frame(top_genus)


abund_df <- abund_df %>%
  dplyr::mutate(pick_genus = ifelse(Genus %in% top_genus,
                                    yes = Genus,
                                    "Others"))

# Plot
abund_df$pick_genus <- as.factor(abund_df$pick_genus)
abund_df$pick_genus <- factor(abund_df$pick_genus,
  levels = c(top_genus, "Others")
)

abund_df <- abund_df[order(abund_df$host.age, decreasing = FALSE), ]

abund_plot <-
  ggplot(abund_df, aes(x = Sam_rep, y = Abundance, fill = pick_Phylum)) +
  geom_bar(position = "fill", stat = "identity") +
  theme_classic() +
  scale_y_continuous(expand = c(0, 0)) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5))
```

```{r}
## update on heatmap plot functions
make_hcb <- function(data, var, name = NULL, fill_scale = NULL, ...) {
  hcb <- ggplot(data = data, aes_string(x = "index", y = 1, fill = var)) +
    geom_raster(show.legend = TRUE) +
    scale_y_continuous(expand = c(0, 0), breaks = 1, labels = name) +
    scale_x_continuous(expand = c(0, 0)) +
    xlab(NULL) + ylab(NULL) +
    theme(axis.title = element_blank(), axis.ticks = element_blank()) +
    theme(axis.text.x = element_blank()) +
    theme(axis.text.y = element_text(size = 12, face = "bold")) +
    theme(plot.margin = unit(c(0, 0, 0, 0), "lines")) +
    guides(fill = FALSE)
  if (!is.null(fill_scale)) hcb <- hcb + fill_scale
  return(hcb)
}

mush <- function(hmap, hcbs) {
  require(gtable)
  require(grid)
  require(gridExtra)
  cbgs <- lapply(hcbs, ggplotGrob)
  hmg <- ggplotGrob(hmap)
  # Make sure both plots have the same dimensions in grob objects
  for (i in seq_along(cbgs)) {
    cbgs[[i]] <- gtable_add_cols(cbgs[[i]], widths = unit(1, "null"), pos = 8)
    cbgs[[i]] <- gtable_add_cols(cbgs[[i]], widths = unit(1, "null"), pos = 8)
  }
  ## Make sure that both plots have the same widths
  cbwidths <- lapply(cbgs, function(x) x$widths)
  maxwidth <- do.call(unit.pmax, cbwidths)
  maxwidth <- unit.pmax(hmg$widths, maxwidth)
  ## replace widths in each grob object with maxWidth
  hmg$widths <- maxwidth
  for (i in seq_along(cbgs)){
    cbgs[[i]]$widths <- maxwidth
  }
  heights <- unit.c(unit(rep(1, length(cbgs)), "lines"), unit(1, "null"))
  rval <- do.call(arrangeGrob,
                  args = c(cbgs,
                           list(hmg),
                           ncol = 1,
                           heights = list(heights)))
  return(rval)
}
```

# Filter phyloseq
```{r}
ps_1 <- ps_filter(all.fil, associated.phenotype != "D000236") # Health vs CRC
ps_2 <- ps_filter(all.fil, associated.phenotype != "D015179") # Health vs Ad
ps_3 <- ps_filter(all.fil, associated.phenotype != "D006262") # CRC vs Ad
```

# Differential abundance

```{r}
mas_1 <- Maaslin2(
  input_data = data.frame(otu_table(ps_1)),
  input_metadata = data.frame(sample_data(ps_1)),
  output = "output/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_1 <- mas_1$results

fdr_mas_1 <- dfmas_res_1

fdr_mas_1$Species <-  phyloseq::tax_table(ps_1)[fdr_mas_1$feature, "Species"]

dim(fdr_mas_1) #307 species
head(fdr_mas_1[order(fdr_mas_1$coef, decreasing = TRUE), ], n = 50)

fc_mas_1 <- exp(fdr_mas_1$coef) ## Antilog coef
fdr_mas_1$log2FoldChange <- log2(fc_mas_1)
rownames(fdr_mas_1) <- fdr_mas_1$feature

fdr_mas_1$Species <- as.character(unclass(fdr_mas_1$Species))
```

# ANCOMBC
```{r}
ps_1
ps_1_keep <- prune_taxa(rowSums(abundances(ps_1) > 0) >= 17,
  ps_1
) #at least 17 samples = 10% of the pool
summary(sample_sums(ps_1_keep) / sample_sums(ps_1)) #99.73% data is covered

#----------#
#ANCOMBC
ancom_1 <- ancombc(data = ps_1_keep,
                   formula = "run.ID + associated.phenotype",
                   tax_level = NULL,
                   p_adj_method = "fdr",
                   lib_cut = 1000,
                   prv_cut = 0.001,
                   group = "associated.phenotype",
                   struc_zero = TRUE,
                   neg_lb = TRUE,
                   tol = 1e-05,
                   alpha = 0.05,
                   global = TRUE)

## consider conserve=TRUE -> "sample size > 30" = large
```

# Visualization
```{r}
volcano_1 <- ggplot(fdr_mas_1, aes(x = log2FoldChange,
                                   y = -log10(qval),
                                   Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_1, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_1, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_1, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_1, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_1,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2.html")
```
# Filter by studyID

## PRJEB6070
```{r}
ps_6070 <- ps_filter(ps_1, project.ID == "PRJEB6070")

# maaslin2

mas_6070 <- Maaslin2(
  input_data = data.frame(otu_table(ps_6070)),
  input_metadata = data.frame(sample_data(ps_6070)),
  output = "output_PRJEB6070/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

# Ancombc2
ancom_6070 <- ancombc2(data = ps_6070, assay_name = "counts",
                       tax_level = "ASV",
                       fix_formula = "associated.phenotype",
                       rand_formula = NULL,
                       p_adj_method = "fdr", pseudo_sens = TRUE,
                       prv_cut = 0.001, lib_cut = 100, s0_perc = 0.05,
                       group = "associated.phenotype",
                       struc_zero = TRUE, neg_lb = TRUE,
                       alpha = 0.05, n_cl = 10, verbose = TRUE,
                       global = FALSE, pairwise = FALSE,
                       dunnet = FALSE, trend = FALSE,
                       iter_control = list(tol = 1e-2, max_iter = 100,
                                           verbose = TRUE),
                       em_control = list(tol = 1e-5, max_iter = 100),
                       lme_control = lme4::lmerControl(),
                       mdfdr_control = list(fwer_ctrl_method = "fdr", B = 100),
                       trend_control = list(contrast = list(matrix(c(1, 0, -1, 1),
                                                                   nrow = 2,
                                                                   byrow = TRUE),
                                                             matrix(c(-1, 0, 1, -1),
                                                                    nrow = 2,
                                                                    byrow = TRUE),
                                                             matrix(c(1, 0, 1, -1),
                                                                    nrow = 2,
                                                                    byrow = TRUE)),
                                            node = list(2, 2, 1),
                                            solver = "ECOS",
                                            B = 10))

tab_zero <- ancom_6070$zero_ind
tab_zero %>%
  datatable(caption = "The detection of structural zeros")

res_prim <- ancom_6070$res
res_prim %>%
  datatable(caption = "Result ANCOMBC2")
write.csv(res_prim, file = "ancombc.csv")

###

dfmas_res_6070 <- mas_6070$results
fdr_mas_6070 <- dfmas_res_6070

fdr_mas_6070$Species <- tax_table(ps_6070)[fdr_mas_6070$feature, "Species"]
fdr_mas_6070$Genus <- tax_table(ps_6070)[fdr_mas_6070$feature, "Genus"]

dim(fdr_mas_6070) #591 species
head(fdr_mas_6070[order(fdr_mas_6070$coef, decreasing = FALSE), ], n = 50)

fc_mas_6070 <- exp(fdr_mas_6070$coef) ## Antilog coef
fdr_mas_6070$log2FoldChange <- log2(fc_mas_6070)
rownames(fdr_mas_6070) <- fdr_mas_6070$feature

fdr_mas_6070$Species <- as.character(unclass(fdr_mas_6070$Species))
write.csv(fdr_mas_6070, "output_PRJEB6070/species.csv")

volcano_6070 <- ggplot(fdr_mas_6070, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_6070, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_6070, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_6070, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_6070, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_6070,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_6070.html")
```

## PRJEB27928
```{r}
ps_7928 <- ps_filter(ps_1, project.ID == "PRJEB27928")

# maaslin2

mas_7928 <- Maaslin2(
  input_data = data.frame(otu_table(ps_7928)),
  input_metadata = data.frame(sample_data(ps_7928)),
  output = "output_PRJEB27928/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

# Ancombc2
ancom_7928 <- ancombc2(data = ps_7928, assay_name = "counts", tax_level = "Species",
                       fix_formula = "associated.phenotype",
                       rand_formula = NULL,
                       p_adj_method = "fdr", pseudo_sens = TRUE,
                       prv_cut = 0.001, lib_cut = 1000, s0_perc = 0.05,
                       group = "associated.phenotype",
                       struc_zero = TRUE, neg_lb = TRUE,
                       alpha = 0.05, n_cl = 2, verbose = TRUE,
                       global = TRUE, pairwise = TRUE, dunnet = TRUE, trend = TRUE,
                       iter_control = list(tol = 1e-2, max_iter = 20,
                                           verbose = TRUE),
                       em_control = list(tol = 1e-5, max_iter = 100),
                       lme_control = lme4::lmerControl(),
                       mdfdr_control = list(fwer_ctrl_method = "holm", B = 100),
                       trend_control = list(contrast = list(matrix(c(1, 0, -1, 1),
                                                                   nrow = 2,
                                                                   byrow = TRUE),
                                                             matrix(c(-1, 0, 1, -1),
                                                                  nrow = 2,
                                                                   byrow = TRUE),
                                                             matrix(c(1, 0, 1, -1),
                                                                  nrow = 2,
                                                                   byrow = TRUE)),
                                            node = list(2, 2, 1),
                                            solver = "ECOS",
                                            B = 10))

tab_zero <- ancom_6070$zero_ind
tab_zero %>%
datatable(caption = "The detection of structural zeros")

res_prim <- output$res
res_prim %>%
  datatable(caption = "Result ANCOMBC2")
write.csv(res_prim, file = "ancombc.csv")

dfmas_res_7928 <- mas_7928$results
fdr_mas_7928 <- dfmas_res_7928

fdr_mas_7928$Species <- tax_table(ps_7928)[fdr_mas_7928$feature, "Species"]
fdr_mas_7928$Genus <- tax_table(ps_7928)[fdr_mas_7928$feature, "Genus"]

dim(fdr_mas_7928) #879 species
head(fdr_mas_7928[order(fdr_mas_7928$coef, decreasing = FALSE), ], n = 50)

fc_mas_7928 <- exp(fdr_mas_7928$coef) ## Antilog coef
fdr_mas_7928$log2FoldChange <- log2(fc_mas_7928)
rownames(fdr_mas_7928) <- fdr_mas_7928$feature

fdr_mas_7928$Species <- as.character(unclass(fdr_mas_7928$Species))
write.csv(fdr_mas_7928, "output_PRJEB27928/species.csv")

volcano_7928 <- ggplot(fdr_mas_7928, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_7928, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_7928, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_7928, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_7928, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_7928,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_7928.html")
```

## PRJEB7774
```{r}
ps_7774 <- ps_filter(ps_1, project.ID == "PRJEB7774")

# maaslin2

mas_7774 <- Maaslin2(
  input_data = data.frame(otu_table(ps_7774)),
  input_metadata = data.frame(sample_data(ps_7774)),
  output = "output_PRJEB7774/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_7774 <- mas_7774$results
fdr_mas_7774 <- dfmas_res_7774

fdr_mas_7774$Species <- tax_table(ps_7774)[fdr_mas_7774$feature, "Species"]
fdr_mas_7774$Genus <- tax_table(ps_7774)[fdr_mas_7774$feature, "Genus"]

dim(fdr_mas_7774) #883 species
head(fdr_mas_7774[order(fdr_mas_7774$coef, decreasing = FALSE), ], n = 50)

fc_mas_7774 <- exp(fdr_mas_7774$coef) ## Antilog coef
fdr_mas_7774$log2FoldChange <- log2(fc_mas_7774)
rownames(fdr_mas_7774) <- fdr_mas_7774$feature

fdr_mas_7774$Species <- as.character(unclass(fdr_mas_7774$Species))
write.csv(fdr_mas_7774, "output_PRJEB7774/species.csv")

volcano_7774 <- ggplot(fdr_mas_7774, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_7774, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_7774, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_7774, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_7774, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_7774,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_7774.html")
```
## PRJEB10878
```{r}
ps_0878 <- ps_filter(ps_1, project.ID == "PRJEB10878")

# maaslin2

mas_0878 <- Maaslin2(
  input_data = data.frame(otu_table(ps_0878)),
  input_metadata = data.frame(sample_data(ps_0878)),
  output = "output_PRJEB10878/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_0878 <- mas_0878$results
fdr_mas_0878 <- dfmas_res_0878

fdr_mas_0878$Species <- tax_table(ps_0878)[fdr_mas_0878$feature, "Species"]
fdr_mas_0878$Genus <- tax_table(ps_0878)[fdr_mas_0878$feature, "Genus"]

dim(fdr_mas_0878) #883 species
head(fdr_mas_0878[order(fdr_mas_0878$coef, decreasing = FALSE), ], n = 50)

fc_mas_0878 <- exp(fdr_mas_0878$coef) ## Antilog coef
fdr_mas_0878$log2FoldChange <- log2(fc_mas_0878)
rownames(fdr_mas_0878) <- fdr_mas_0878$feature

fdr_mas_0878$Species <- as.character(unclass(fdr_mas_0878$Species))
write.csv(fdr_mas_0878, "output_PRJEB10878/species.csv")

volcano_0878 <- ggplot(fdr_mas_0878, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_0878, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_0878, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_0878, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_0878, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_0878,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_0878.html")
```
## PRJDB4176
```{r}
ps_4176 <- ps_filter(ps_1, project.ID == "PRJDB4176")

# maaslin2

mas_4176 <- Maaslin2(
  input_data = data.frame(otu_table(ps_4176)),
  input_metadata = data.frame(sample_data(ps_4176)),
  output = "output_PRJDB4176/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_4176 <- mas_4176$results
fdr_mas_4176 <- dfmas_res_4176

fdr_mas_4176$Species <- tax_table(ps_4176)[fdr_mas_4176$feature, "Species"]
fdr_mas_4176$Genus <- tax_table(ps_4176)[fdr_mas_4176$feature, "Genus"]

dim(fdr_mas_4176) #883 species
head(fdr_mas_4176[order(fdr_mas_4176$coef, decreasing = FALSE), ], n = 50)

fc_mas_4176 <- exp(fdr_mas_4176$coef) ## Antilog coef
fdr_mas_4176$log2FoldChange <- log2(fc_mas_4176)
rownames(fdr_mas_4176) <- fdr_mas_4176$feature

fdr_mas_4176$Species <- as.character(unclass(fdr_mas_4176$Species))
write.csv(fdr_mas_4176, "output_PRJEB4176/species.csv")

volcano_4176 <- ggplot(fdr_mas_4176, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_4176, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_4176, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_4176, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_4176, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_4176,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_4176.html")
```

## PRJEB12449
```{r}
ps_2449 <- ps_filter(ps_1, project.ID == "PRJEB12449")

# maaslin2

mas_2449 <- Maaslin2(
  input_data = data.frame(otu_table(ps_2449)),
  input_metadata = data.frame(sample_data(ps_2449)),
  output = "output_PRJEB12449/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_2449 <- mas_2449$results
fdr_mas_2449 <- dfmas_res_2449

fdr_mas_2449$Species <- tax_table(ps_2449)[fdr_mas_2449$feature, "Species"]
fdr_mas_2449$Genus <- tax_table(ps_2449)[fdr_mas_2449$feature, "Genus"]

dim(fdr_mas_2449) #883 species
head(fdr_mas_2449[order(fdr_mas_2449$coef, decreasing = FALSE), ], n = 50)

fc_mas_2449 <- exp(fdr_mas_2449$coef) ## Antilog coef
fdr_mas_2449$log2FoldChange <- log2(fc_mas_2449)
rownames(fdr_mas_2449) <- fdr_mas_2449$feature

fdr_mas_2449$Species <- as.character(unclass(fdr_mas_2449$Species))
write.csv(fdr_mas_2449, "output_PRJEB12449/species.csv")

volcano_2449 <- ggplot(fdr_mas_2449, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_2449, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_2449, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_2449, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_2449, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_2449,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_2449.html")
```
## PRJNA447983
```{r}
ps_7983 <- ps_filter(ps_1, project.ID == "PRJNA447983")

# maaslin2

mas_7983 <- Maaslin2(
  input_data = data.frame(otu_table(ps_7983)),
  input_metadata = data.frame(sample_data(ps_7983)),
  output = "output_PRJNA447983/",
  min_abundance = 0,
  min_prevalence = 0,
  normalization = "NONE",
  transform = "LOG",
  analysis_method = "LM",
  max_significance = 0.05,
  fixed_effects = "associated.phenotype",
  random_effects = c("host.age", "BMI"),
  reference = "run.ID",
  plot_heatmap = FALSE,
  plot_scatter = FALSE,
  standardize = FALSE,
  cores = 22
)

dfmas_res_7983 <- mas_7983$results
fdr_mas_7983 <- dfmas_res_7983

fdr_mas_7983$Species <- tax_table(ps_7983)[fdr_mas_7983$feature, "Species"]
fdr_mas_7983$Genus <- tax_table(ps_7983)[fdr_mas_7983$feature, "Genus"]

dim(fdr_mas_7983) #883 species
head(fdr_mas_7983[order(fdr_mas_7983$coef, decreasing = FALSE), ], n = 50)

fc_mas_7983 <- exp(fdr_mas_7983$coef) ## Antilog coef
fdr_mas_7983$log2FoldChange <- log2(fc_mas_7983)
rownames(fdr_mas_7983) <- fdr_mas_7983$feature

fdr_mas_7983$Species <- as.character(unclass(fdr_mas_7983$Species))
write.csv(fdr_mas_7983, "output_PRJNA447983/species.csv")

volcano_7983 <- ggplot(fdr_mas_7983, aes(x = log2FoldChange,
                                         y = -log10(qval),
                                         Species = Species)) +
  geom_hline(aes(yintercept = max(-log10(qval))),
             linetype = "dashed",
             color = "grey30") +
  geom_point(data = subset(fdr_mas_7983, qval > 0.05),
             size = 3, color = "grey30") +
  geom_point(data = subset(fdr_mas_7983, qval <= 0.05 & log2FoldChange < 0),
             size = 3, color = "#457b9d") +
  geom_point(data = subset(fdr_mas_7983, qval <= 0.05 & log2FoldChange > 0),
             size = 3, color = "#d62828") +
  geom_hline(yintercept = -log10(0.05), col = "grey30",
             linetype = 1) +
  geom_vline(xintercept = 0, col = "grey30",
             linetype = 1) +
  geom_text(aes(x = 3, y = -log10(0.03), label = "q-value = 0.05"),
            hjust = 1, vjust = 0,
            color = "grey30",
            size = 5) +
  # scale_y_continuous(expand = c(0, 0), limits = c(0, 15)) +
  scale_x_continuous(expand = c(0, 0), limits = c(-4, 4))  +
  theme_classic() +
  labs(title = "Differential abundance",
       subtitle = "Healthy (left) vs CRC (right)") +
  theme(plot.title = element_text(size = 10, face = "bold",
                                  vjust = 0, hjust = 0),
        plot.subtitle = element_text(size = 10,
                                     color = "grey50",
                                     vjust = 0, hjust = 0),
        axis.title.x = element_text(vjust = 0, hjust = 0.5, size = 10),
        axis.text.x = element_text(size = 10),
        axis.title.y = element_text(size = 10),
        axis.text.y = element_text(size = 10),
        axis.line.y.left = element_blank()) +
  geom_segment(aes(x = -4, y = -Inf,
                   xend = -4, yend = max(-log10(qval))),
               size = 1, linetype = "solid") +
  geom_segment(aes(x = -4, y = max(-log10(qval)),
                   xend = -4, yend = Inf),
               size = 1, linetype = "dotted")

ggplotly(volcano_7983, tooltip = c("log2FoldChange", "-log10(qval)", "Species"))
saveWidget(ggplotly(volcano_7983,
                    tooltip = c("log2FoldChange",
                                "-log10(qval)",
                                "Species")),
           file = "maaslin2_7983.html")
```

## Summary DA:
```{r}
fdr_mas_6070
fdr_mas_7928
fdr_mas_7774
fdr_mas_4176
fdr_mas_0878
fdr_mas_2449
fdr_mas_7983

fdr_mas_6070$study_ID <- "PRJEB6070"
fdr_mas_7928$study_ID <- "PRJEB27928"
fdr_mas_7774$study_ID <- "PRJEB7774"
fdr_mas_0878$study_ID <- "PRJEB10878"
fdr_mas_4176$study_ID <- "PRJEB4176"
fdr_mas_2449$study_ID <- "PRJEB12449"
fdr_mas_7983$study_ID <- "PRJEB27928"

rownames(fdr_mas_6070) <- NULL
rownames(fdr_mas_7928) <- NULL
rownames(fdr_mas_7774) <- NULL
rownames(fdr_mas_0878) <- NULL
rownames(fdr_mas_4176) <- NULL
rownames(fdr_mas_2449) <- NULL
rownames(fdr_mas_7983) <- NULL

combine_df <- rbind(fdr_mas_6070,
                    fdr_mas_7928,
                    fdr_mas_7774,
                    fdr_mas_0878,
                    fdr_mas_4176,
                    fdr_mas_2449,
                    fdr_mas_7983)
combine_df$study_ID <- as.factor(combine_df$study_ID)

# filter significant biomarker
significant_df <- subset(combine_df, qval <= 0.05)


# Create a new variable for binned log2FoldChange values
significant_df$log2FoldChange_binned <- cut(significant_df$log2FoldChange,
                                            breaks = c(-6, -5, -4, -3, -2, -1,
                                                       0,
                                                       1, 2, 3, 4, 5, 6))

# Create the plot
biomarker_plot <- ggplot(significant_df,
                         aes(x = Species,
                             y = study_ID,
                             fill = log2FoldChange_binned,
                             log2FoldChange = log2FoldChange,
                             qval = qval)) +
  geom_tile(linejoin = "round") +
  scale_fill_manual(values = c("#03045e",
                               "#023e8a",
                               "#0077b6",
                               "#0096c7",
                               "#00b4d8",
                               "#90e0ef",
                               "#ffba08",
                               "#f48c06",
                               "#dc2f02",
                               "#9d0208",
                               "#6a040f",
                               "#370617"),
                    breaks = levels(significant_df$log2FoldChange_binned)) +
  coord_fixed(1) +
  theme_classic() +
  theme(axis.text.x = element_text(face = "italic",
                                   angle = 45))
ggplotly(biomarker_plot)
saveWidget((ggplotly(biomarker_plot,
                     tooltip = c("Species",
                                 "log2FoldChange",
                                 "qval"))),
           file = "biomarker.html")
```

# plot network
```{r}
##################################################################
#-----------------------------------------------------------------
## Taken from CCLASSO methodology
# The CCLasso function from github
################################################################################
# File: cclasso.R
# Aim : Correlation inference for compositional data through lasso
#-------------------------------------------------------------------------------
# Author : Fang Huaying (Peking University)
# Email  : hyfang@pku.edu.cn
# Date   : 2016-01-08
# Version: 2.0
#-------------------------------------------------------------------------------
# Main function: cclasso(x, counts = FALSE, pseudo = 0.5, k_cv = 3,
#                        lam_int = c(1e-4, 1), k_max = 20, n_boot = 20)
#
#  Input:
#           x ------ n x p data matrix (row/column is sample/variable)
#                    n samples & p compositional variables
#      counts ------ Is the compositional data matrix a count matrix?
#                    Default: FALSE
#      pseudo ------ pseudo count if counts = TRUE
#                    Default: 0.5
#        k_cv ------ folds of cross validation
#                    Default: 3
#     lam_int ------ tuning parameter interval
#                    Default: [1e-4, 1]
#       k_max ------ maximum iterations for golden section method
#                    Default: 20
#      n_boot ------ Bootstrap times
#                    Default: 20
#  Output:
#      A list structure contains:
#       var_w ------ variance estimation
#       cor_w ------ correlation estimation
#      p_vals ------ p-values for elements of cor_w equal 0 or not
#      lambda ------ final tuning parameter
#     info_cv ------ information for cross validation
#-------------------------------------------------------------------------------
#CClasso function
cclasso <- function(x, counts = FALSE, pseudo = 0.5, k_cv = 3,
                    lam_int = c(1e-4, 1), k_max = 20, n_boot = 20) {
  p <- ncol(x)

  if (counts) {
    x <- x + pseudo
    x <- x / rowSums(x)
  }
  x <- log(x)
  vx2 <- stats::var(x)

  # Diagonal weight for loss function
  rmean_vx2 <- rowMeans(vx2)
  wd <- 1 / diag(vx2 - rmean_vx2 - rep(rmean_vx2, each = p) + mean(rmean_vx2))
  wd2 <- sqrt(wd)
  # Some global parameters for optimization with single lambda
  rho <- 1
  u_f <- eigen(diag(p) - 1 / p)$vectors
  wd_u <- (t(u_f) %*% (wd * u_f))[-p, -p]
  wd_u_eig <- eigen(wd_u)
  d0_wd <- 1 / ((rep(wd_u_eig$values, each = p - 1) + wd_u_eig$values) /
                  (2 * rho) + 1)
  u0_wd <- wd_u_eig$vectors

  # Golden section method for the selection of lambda (log10 scale)
  sigma <- vx2
  lam_int2 <- log10(range(lam_int))
  a1 <- lam_int2[1]
  b1 <- lam_int2[2]
  # Store lambda and corresponding cross validation's loss
  lams <- NULL
  fvals <- NULL
  # Two trial points in first
  a2 <- a1 + 0.382 * (b1 - a1)
  b2 <- a1 + 0.618 * (b1 - a1)
  fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv,
                         sigma = sigma,
                         wd = wd,
                         u_f = u_f,
                         u0_wd = u0_wd,
                         d0_wd = d0_wd,
                         wd2 = wd2)
  lams <- c(lams, b2)
  fvals <- c(fvals, fb2$cv_loss)
  fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv,
                         sigma = fb2$sigma,
                         wd = wd,
                         u_f = u_f,
                         u0_wd = u0_wd,
                         d0_wd = d0_wd,
                         wd2 = wd2)
  lams <- c(lams, a2)
  fvals <- c(fvals, fa2$cv_loss)
  # Error tolerance for convergence
  err_lam2 <- 1e-1 * max(1, lam_int2)
  err_fval <- 1e-4
  err <- b1 - a1
  k <- 0
  while (err > err_lam2 && k < k_max) {
    fval_max <- max(fa2$cv_loss, fb2$cv_loss)

    if (fa2$cv_loss > fb2$cv_loss) {
      a1 <- a2
      a2 <- b2
      fa2 <- fb2
      b2 <- a1 + 0.618 * (b1 - a1)
      fb2 <- cv_loss_cclasso(lambda2 = 10^b2 / rho, x = x, k_cv = k_cv,
                             sigma = fa2$sigma,
                             wd = wd,
                             u_f = u_f,
                             u0_wd = u0_wd,
                             d0_wd = d0_wd,
                             wd2 = wd2)
      lams <- c(lams, b2)
      fvals <- c(fvals, fb2$cv_loss)
    } else {
      b1 <- b2
      b2 <- a2
      fb2 <- fa2
      a2 <- a1 + 0.382 * (b1 - a1)
      fa2 <- cv_loss_cclasso(lambda2 = 10^a2 / rho, x = x, k_cv = k_cv,
                             sigma = fb2$sigma,
                             wd = wd,
                             u_f = u_f,
                             u0_wd = u0_wd,
                             d0_wd = d0_wd,
                             wd2 = wd2)

      lams <- c(lams, a2)
      fvals <- c(fvals, fa2$cv_loss)
    }
    fval_min <- min(fa2$cv_loss, fb2$cv_loss)

    k <- k + 1
    err <- b1 - a1
    if (abs(fval_max - fval_min) / (1 + fval_min) <= err_fval) {
      break
    }
  }
  info_cv <- list(lams = lams, fvals = fvals, k = k + 2,
                  lam_int = 10^c(a1, b1))
  if (a1 == lam_int2[1] || b1 == lam_int2[2]) {
    cat("WARNING:\n", "\tOptimal lambda is near boundary! ([", 10^a1, ",",
        10^b1, "])\n", sep = "")
  }

  lambda <- 10^((a2 + b2) / 2)
  # Bootstrap for cclasso
  lambda2 <- lambda / rho
  info_boot <- boot_cclasso(x = x, sigma = fb2$sigma, lambda2 = lambda2,
                            n_boot = n_boot,
                            wd = wd,
                            u_f = u_f,
                            u0_wd = u0_wd,
                            d0_wd = d0_wd)

  return(list(var_w = info_boot$var_w, cor_w = info_boot$cor_w,
              p_vals = info_boot$p_vals,
              lambda = lambda,
              info_cv = info_cv))
}

## Bootstrap for cclasso
boot_cclasso <- function(x, sigma, lambda2, n_boot = 20,
                         wd, u_f, u0_wd, d0_wd) {
  n <- nrow(x)
  p <- ncol(x)

  # Store the result of bootstrap
  cors_boot <- matrix(0, nrow = p * (p - 1) / 2, ncol = n_boot + 1)
  vars_boot <- matrix(0, nrow = p, ncol = n_boot + 1)
  cors_mat <- matrix(0, p, p)
  ind_low <- lower.tri(cors_mat)

  # Bootstrap procedure
  sam_boot <- matrix(sample(1:n, size = n * n_boot, replace = TRUE),
                     ncol = n_boot)
  for (k in 1:n_boot) {
    ind_samp <- sam_boot[, k]
    sigma2 <- cclasso_sub(sigma = sigma, vx = var(x[ind_samp, ]),
                          lambda2 = lambda2,
                          wd = wd,
                          u_f = u_f,
                          u0_wd = u0_wd,
                          d0_wd = d0_wd)
    vars_boot[, k] <- diag(sigma2)
    is <- 1 / sqrt(vars_boot[, k])
    cors_mat <- is * sigma2 * rep(is, each = p)
    cors_boot[, k] <- cors_mat[ind_low]
  }
  is <- 1 / sqrt(diag(sigma))
  cors_mat <- sigma * is * rep(is, each = p)
  cors_boot[, n_boot + 1] <- cors_mat[ind_low]
  vars_boot[, n_boot + 1] <- diag(sigma)

  #----------------------------------------
  # Variance estimation via bootstrap
  vars2 <- rowMeans(vars_boot)
  #----------------------------------------
  # Correlations' relationship for artificial null sample
  tol_cor <- 1e-3
  sam_art0 <- matrix(rnorm(n * p), nrow = n) * rep(sqrt(vars2), each = n)
  cors_art0 <- cor(sam_art0)[ind_low]
  sam_art <- sam_art0 - log(rowSums(exp(sam_art0)))
  sigma_art <- cclasso_sub(sigma = sigma, vx = var(sam_art),
                           lambda2 = lambda2,
                           wd = wd,
                           u_f = u_f,
                           u0_wd = u0_wd,
                           d0_wd = d0_wd)
  is <- 1 / sqrt(diag(sigma_art))
  cors_mat <- is * sigma_art * rep(is, each = p)
  cors_art2 <- cors_mat[ind_low]
  # Bias of estimation between absolute data and cclasso of compositional data
  cors0m2 <- log(((1 + cors_art0) * (1 - cors_art2)) /
                   ((1 + cors_art2) * (1 - cors_art0)))
  tmp <- abs(cors_art2) >= tol_cor
  bias02 <- ifelse(sum(tmp), median(abs(cors0m2)[tmp]), 0)
  # Modification of estimation for cclasso
  cors2 <- log((1 + cors_boot) / (1 - cors_boot))
  cors2mod <- (cors_boot >= tol_cor) * (cors2 + bias02) +
    (cors_boot <= -tol_cor) * (cors2 - bias02)
  cors2mod <- 1 - rowMeans(2 / (exp(cors2mod) + 1))
  cors2_mat <- diag(p)
  cors2_mat[ind_low] <- cors2mod
  cors2_mat <- t(cors2_mat)
  cors2_mat[ind_low] <- cors2mod
  # P-values with null distribution of correlation estimations of absolute data
  p_vals <- pt(cors2mod * sqrt((n - 2) / (1 - cors2mod^2)), df = n - 2)
  p_vals <- ifelse(p_vals <= 0.5, p_vals, 1 - p_vals)
  pval_mat <- diag(p)
  pval_mat[ind_low] <- p_vals
  pval_mat <- t(pval_mat)
  pval_mat[ind_low] <- p_vals
  #----------------------------------------

  return(list(var_w = vars2, cor_w = cors2_mat, p_vals = pval_mat))
}

# cross validation's loss of cclasso for single lambda
cv_loss_cclasso <- function(lambda2, x, k_cv, sigma,
                            wd, u_f, u0_wd, d0_wd, wd2) {
  n <- nrow(x)
  p <- ncol(x)

  n_b <- floor(n / k_cv)
  cv_loss <- 0
  for (k in 1:k_cv) {
    itest <- (n_b * (k - 1) + 1):(n_b * k)
    vxk <- stats::var(x[itest, ])
    vx2k <- stats::var(x[-itest, ])

    sigma <- cclasso_sub(sigma = sigma, vx = vx2k, lambda2 = lambda2,
                         wd = wd,
                         u_f = u_f,
                         u0_wd = u0_wd,
                         d0_wd = d0_wd)
    dsig <- sigma - vxk
    tmp <- rowMeans(dsig)
    dsig <- dsig - tmp - rep(tmp, each = p) + mean(tmp)
    cv_loss <- cv_loss + base::norm(wd2 * dsig, "F")^2
  }

  return(list(cv_loss = cv_loss, sigma = sigma))
}
#-------------------------------------------------------------------------------
# cclasso for single lambda
cclasso_sub <- function(sigma, vx, lambda2,
                        wd, u_f, u0_wd, d0_wd,
                        k_max = 200, x_tol = 1e-4) {
  p <- ncol(sigma)
  sigma2 <- sigma
  lamb_mat <- matrix(0, p, p)
  lambda2 <- matrix(lambda2, p, p)
  diag(lambda2) <- 0

  k <- 0
  err <- 1
  while (err > x_tol && k < k_max) {
    # Update sigma
    x_sigma <- t(u_f) %*% ((sigma2 - vx) - lamb_mat) %*% u_f
    x_sigma[-p, -p] <- u0_wd %*% ((t(u0_wd) %*% x_sigma[-p, -p] %*% u0_wd) *
                                    d0_wd) %*% t(u0_wd)
    sigma_new <- vx + u_f %*% x_sigma %*% t(u_f)
    # Update sigma2
    a <- lamb_mat + sigma_new
    sigma2_new <- (a > lambda2) * (a - lambda2) +
      (a < -lambda2) * (a + lambda2)
    # Update Lambda
    lamb_mat <- lamb_mat + (sigma_new - sigma2_new)

    err <- max(abs(sigma_new - sigma) / (abs(sigma) + 1),
               abs(sigma2_new - sigma2) / (abs(sigma2) + 1))
    k <- k + 1
    sigma <- sigma_new
    sigma2 <- sigma2_new
  }

  if (k >= k_max) {
    cat("WARNING of cclasso_sub:\n", "\tMaximum Iteration:",
        k_max,
        "&& Relative error:", err, "!\n")
  }
  return(sigma)
}
#------------------------------------------------------------------------------
#------------------------------------------------------------------------------
plot_cclasso <- function(cor_map, adj_mat, phyloseq) {
  # Loading library
  require(igraph)
  require(plyr)
  require(RColorBrewer)
  require(cols4all)
  # create a graph object from the adjacency matrix
  g <- graph.adjacency(adj_mat)
  # get the edges as well as the original
  edgelist <- get.edgelist(g)
  edge_cor <- apply(edgelist,
                    MARGIN = 1,
                    FUN = function(x) {
                      cor_map[x[1], x[2]]
                    })
  edgelist <- data.frame(edgelist, edge_cor)
  colnames(edgelist) <- c("from", "to", "cor")
  # filter edges that link the same node
  filt <- edgelist$from == edgelist$to
  print("There are some edges that link the same node... now deleting...")
  edgelist <- edgelist[!filt, ]
  edgelist[, 1:2] <- t(apply(edgelist,
                             MARGIN = 1,
                             FUN = function(x) {
                               sortedges(x[1:2])
                             }))
  dupl <- duplicated(paste(edgelist$from, edgelist$to))
  sum(dupl)
  edgelist <- edgelist[!dupl, ]
  # Take only one direction

  # get the annotation for the nodes that are implicated in the network.
  all.nodes <- unique(c(as.character(edgelist$from), as.character(edgelist$to)))
  # Get the taxonomic classification and abundance of OTUs from original data.
  feat_annot <- taxa_names(phyloseq)[as.numeric(all.nodes)] #OTU names
  feat_annot <- as.data.frame(feat_annot)
  rownames(feat_annot) <- feat_annot$feat_annot
  feat_annot$name <- all.nodes
  feat_annot <- feat_annot[, c(2, 1)]
  feat_annot$size <- as.vector(
    taxa_sums(phyloseq)[as.numeric(all.nodes)]
  ) # abundance
  feat_annot$Family <- as.vector(
    tax_table(phyloseq)[as.numeric(all.nodes), "Family"]
  ) #Assign taxonomy of the Family level
  print("Finished first part")

  # create the igraph object
  gd <- graph.data.frame(d = edgelist, directed = FALSE, vertices = feat_annot)
  degall <- igraph::degree(gd, v = V(gd), mode = "all")
  V(gd)$name.origin <- tax_table(phyloseq)[as.numeric(all.nodes), "Species"]
  V(gd)$name <- tax_table(phyloseq)[as.numeric(all.nodes), "Species"]
  V(gd)$otu <- rownames(tax_table(phyloseq)[as.numeric(all.nodes), ])
  gd <- set.vertex.attribute(gd, "degree", index = V(gd)$name, value = degall)
  #gd <- set.vertex.attribute(gd, "degree", index = V(gd)$otu, value = degall)

  # Calculate betweenness for all nodes
  betall <- igraph::betweenness(gd,
                                v = V(gd),
                                directed = FALSE) /
    (((vcount(gd) - 1) * (vcount(gd) - 2)) / 2)
  betall_norm <- (betall - min(betall)) /
    (max(betall) - min(betall))
  rm(betall)

  # Add new node/edge attributes based on
  # the calculated node properties/similarities
  gd <- set.vertex.attribute(gd,
                             "betweenness",
                             index = V(gd)$name,
                             value = betall_norm)

  ##Calculate edge properties and add to the network
  #Calculate Dice similarities between all pairs of nodes
  dsall <- similarity.dice(gd, vids = V(gd), mode = "all")
  # The following function will transform a square matrix
  # to an edge driven one and add values to each edge
  f1 <- function(x) {
    data.frame(dice = dsall[which(V(gd)$name == as.character(x$from)),
                            which(V(gd)$name == as.character(x$to))])
  }
  edgelist_ext <- ddply(edgelist, .variables = c("from", "to"),
                        function(x) data.frame(f1(x)))
  dim(edgelist_ext)
  gd <- set.edge.attribute(gd, "similarity", index = E(gd), value = 0)
  E(gd)[as.character(edgelist_ext$from) %--%
          as.character(edgelist_ext$to)]$similarity <- as.numeric(
    edgelist_ext$dice
  )
  # Edge color based on type of correlations
  E(gd)$color <- ifelse(E(gd)$cor < 0,
                        "#EE6677",
                        "#4477AA") # Red for Negative, Blue for Positive
  # Vertex color based on the Order of the OTU
  V(gd)$color <- c(cols4all::c4a("classic20", 19),
                   c4a("okabe", 7),
                   c4a("rainbow", 23))[as.factor(V(gd)$Family)]

  # resume the network
  print(paste("There are", vcount(gd),
              "nodes and", ecount(gd), "edges"))
  summary(gd)

  # compute the layout
  l <- layout.auto(gd)
  par(oma = c(2, 0.2, 0.2, 0.2))

  pl <- plot(gd,
    vertex.label = V(gd)$name,
    vertex.label.cex = 1.2,
    #vertex.size=(log10(V(gd)$size))^1.5/2,
    vertex.size = 5,
    vertex.label.dist = 0.75, vertex.label.color = "black",
    vertex.label.font = 3,
    edge.arrow.size = 2, asp = TRUE,
    #edge.width = abs(E(gd)$cor)*7,
    edge.width = 4,
    rescale = TRUE, layout = l * 1.2
  )
  par(mfrow = c(1, 1),
      fig = c(0, 1, 0, 1),
      oma = c(0, 0, 0, 0),
      mar = c(0, 0, 0, 0),
      new = TRUE)
  ## legend
  legend(1.5,
         1,
         legend = levels(factor(tax_table(phyloseq)[as.numeric(all.nodes),
                                                    "Family"])),
         fill = c(c4a("classic20", 19),
                  c4a("okabe", 7),
                  c4a("rainbow", 23)),
         xpd = TRUE,
         bty = "n",
         cex = 1,
         xjust = 0,
         ncol = 1,
         text.width = 0.5,
         x.intersp = 0.1,
         title = "Family")
  return(pl)
}

## To beautify the cclasso plot
sortedges <- function(var_name = c(node1, node2)) { # nolint
  if (length(var_name) != 2) stop("Two variable names are needed")
  if (var_name[2] < var_name[1]) {
    b <- var_name[1]
    var_name[1] <- var_name[2]
    var_name[2] <- b
  }
  return(var_name)
}
#----------------------------------------------------------#
```
```{r}
ps_6070

#NetCoMi
net_6070 <- netConstruct(data = ps_6070,
                         measure = "spieceasi",
                         normMethod = "none",
                         zeroMethod = "multRepl",
                         sparsMethod = "t-test",
                         dissFunc = "signed",
                         verbose = 3,
                         cores = 16L)

props_net_6070 <- netAnalyze(net_6070,
                             centrLCC = FALSE,
                             clustMethod = "cluster_fast_greedy",
                             hubPar = "eigenvector",
                             normDeg = FALSE,
                             weightDeg = FALSE)

#summary
summary(props_net_6070)

tax_table(ps_6070)[names(sort(props_net_6070$centralities$eigenv1,
                              decreasing = TRUE)[1:10]), ]
tax_table(ps_6070)[props_net_6070$hubs$hubs1, ]

#plot network
net_6070_cormap <- net_6070$assoEst1
colnames(net_6070_cormap) <- NULL
rownames(net_6070_cormap) <- NULL
net_6070_cormap<- as.matrix(net_6070_cormap)

diag(net_6070_cormap) <- 0 ## all self-interactions are set to 0

#Exploration
plot(density(as.numeric(net_6070_cormap)), main="All interactions") ## most interaction has strength centering zero
plot(density(as.numeric(net_6070_cormap[abs(net_6070_cormap)>0.5])), main="filtered")

net_6070_adjmat <- (net_6070_cormap >= 0.9 | net_6070_cormap <= -0.2) + 0

length(which(net_6070_adjmat == 1))

ggplotly(plot_cclasso(net_6070_cormap, net_6070_adjmat, ps_6070))

ggplot2::ggsave(filename = "6070_net.png",
                plot = plot_cclasso(net_6070_cormap,
                                    net_6070_adjmat,
                                    ps_6070),
                device = "png",
                path = "Network/",
                width = 60,
                height = 36,
                units = "in",
                dpi = 400,
                limitsize = FALSE,
                bg = "white")
```